nanoDFT/sparse_symmetric_ERI.py

```diff
@@ -62,35 +80,153 @@ def num_repetitions_fast(value):
+        print("einsum K", time.time()-start)
+        K = np.einsum("ijkl,jk->il", dense_ERI, dm)
+        print("J-K/2", time.time()-start)
+        truth = J - K / 2 * HYB_B3LYP
+
+    print("\n----------")
+    print("nonzero_indices ", time.time()-start)
+    nonzero_indices      = np.nonzero(distinct_ERI)[0].astype(np.int32)
+    print("grap ERI values", time.time()-start)
+    nonzero_distinct_ERI = distinct_ERI[nonzero_indices].astype(np.float32)
+    if not args.skip: 
+        print("dense: ", dense_ERI.shape, dense_ERI.nbytes/10**6)
+    print("sparse: ", distinct_ERI.shape, distinct_ERI.nbytes/10**6)
+    print("sparse+symmetric: ", nonzero_distinct_ERI.shape, nonzero_distinct_ERI.nbytes/10**6)
+    print("per ipu", nonzero_distinct_ERI.nbytes/10**6/16*2) # obs: most time might be spent compiling loading ERI from H->device! 
+    print("----------\n")
+
+    # write code that profiles get_indices/take/segment_sum independently. 
+    if args.prof: 
+        symmetry = 7 
+        dm_indices = get_ijkl(nonzero_indices, symmetry, mask, dm_indices_func_J)  
+
+        size = nonzero_indices.shape[0]
+        total_threads = (1472-1) * 6 
+        remainder = size % total_threads
+        if remainder != 0: 
+            nonzero_indices = np.pad(nonzero_indices, (0, total_threads-remainder))
+        nonzero_indices = nonzero_indices.reshape(total_threads, -1) 
+
+        _dm_indices = np.asarray(_get_ijkl(nonzero_indices, symmetry, N, mask))[:size]
+        assert np.allclose(dm_indices, _dm_indices)
+        exit()
+
+    # Compute number of repetitions per nonzero index and re-scale ERI. 
+    print("compute eri normalization ", time.time()-start)
+    rep = jax.jit(vmap_num_repetitions_fast, backend="cpu")(nonzero_indices)
+    print("perform normalization ", time.time()-start)
+    nonzero_distinct_ERI = nonzero_distinct_ERI / rep
+    mask = np.ones(1)
+    dm = dm.reshape(-1)
+    diff_JK = np.zeros(dm.shape)
+
+    print("pad remainder ", time.time()-start)
+    print(nonzero_distinct_ERI.shape, nonzero_indices.shape)
+    batches = int(args.batches) # perhaps make 10 batches? 
+    remainder = nonzero_indices.shape[0] % (nipu*batches)
+    if remainder != 0:
+        nonzero_indices = np.pad(nonzero_indices, (0,nipu*batches-remainder))
+        nonzero_distinct_ERI = np.pad(nonzero_distinct_ERI, (0,nipu*batches-remainder))
+
+    print("reshape to 16 IPUs pmap", time.time()-start)
+    nonzero_indices = nonzero_indices.reshape(nipu, batches, -1)
+    nonzero_distinct_ERI = nonzero_distinct_ERI.reshape(nipu, batches, -1)
+    print(nonzero_indices.shape)
+    print(nonzero_distinct_ERI.shape)
+
+    print("call pmap", time.time()-start)
+    print("[%i]"%mol.nao_nr())
+    diff_JK = jax.pmap( sparse_symmetric_einsum, in_axes=(0,0,None,None,None, None), static_broadcasted_argnums=(5,), backend=backend, axis_name="p") (nonzero_distinct_ERI, nonzero_indices, dm, mask, diff_JK, args.backend) 
```

Comment:
This line throws an error, I think `diff_JK` can be removed from the arguments, i.e.:

`diff_JK = jax.pmap( sparse_symmetric_einsum, in_axes=(0,0,None,None,None), static_broadcasted_argnums=(4,), backend=backend, axis_name="p") (nonzero_distinct_ERI, nonzero_indices, dm, mask, args.backend) `